modules/imgproc/include/opencv2/imgproc/hal/hal.hpp

```diff
@@ -0,0 +1,53 @@
+#ifndef CV_IMGPROC_HAL_HPP
+#define CV_IMGPROC_HAL_HPP
+
+#include "opencv2/core/cvdef.h"
+#include "opencv2/core/hal/interface.h"
+
+namespace cv { namespace hal {
+
+//! @addtogroup core_hal_functions
+//! @{
+
+struct FilterContext
```

Comment:
I just thought that we could do it in our algorithm-like style:

<pre>
struct Filter2D
{
     virtual ~Filter2D() {}
     virtual void apply(...) = 0;
     static Ptr<Filter2D> create(...);
};
</pre>

what do you think?